File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/event/MeetingBuilder.java
package com.github.dedis.popstellar.model.objects.event;

import com.github.dedis.popstellar.model.objects.Meeting;

import java.util.List;

/** This class is a builder for the Meeting object */
public class MeetingBuilder {
  private String id;
  private String name;
  private long creation;
  private long start;
  private long end;
  private String location;
  private long lastModified;
  private String modificationId;
  private List<String> modificationSignatures;

  public MeetingBuilder() {
    // Empty constructor required
  }

  public MeetingBuilder(Meeting meeting) {
    id = meeting.getId();
    name = meeting.getName();
    creation = meeting.getCreation();
    start = meeting.getStartTimestamp();
    end = meeting.getEndTimestamp();
    location = meeting.getLocation();
    lastModified = meeting.getLastModified();
    modificationId = meeting.getModificationId();
    modificationSignatures = meeting.getModificationSignatures();
  }

  public MeetingBuilder setId(String id) {
    this.id = id;
    return this;
  }

  public MeetingBuilder setName(String name) {
    this.name = name;
    return this;
  }

  public MeetingBuilder setCreation(long creation) {
    this.creation = creation;
    return this;
  }

  public MeetingBuilder setStart(long start) {
    this.start = start;
    return this;
  }

  public MeetingBuilder setEnd(long end) {
    this.end = end;
    return this;
  }

  public MeetingBuilder setLocation(String location) {
    if (location == null) {
      throw new IllegalArgumentException("Location is null");
    }
    this.location = location;
    return this;
  }

  public MeetingBuilder setLastModified(long lastModified) {
    this.lastModified = lastModified;
    return this;
  }

  public MeetingBuilder setModificationId(String modificationId) {
    if (modificationId == null) {
      throw new IllegalArgumentException("Modification Id is null");
    }
    this.modificationId = modificationId;
    return this;
  }

  public MeetingBuilder setModificationSignatures(List<String> modificationSignatures) {
    if (modificationSignatures == null) {
      throw new IllegalArgumentException("Modification signatures list is null");
    }
    this.modificationSignatures = modificationSignatures;
    return this;
  }

  public Meeting build() {
    if (location == null) {
      throw new IllegalStateException("Location is null");
    }
    if (modificationId == null) {
      throw new IllegalArgumentException("Modification Id is null");
    }
    if (modificationSignatures == null) {
      throw new IllegalArgumentException("Modification signatures list is null");
    }
    return new Meeting(
        id,
        name,
        creation,
        start,
        end,
        location,
        lastModified,
        modificationId,
        modificationSignatures);
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/objects/event/MeetingBuilder.kt
package com.github.dedis.popstellar.model.objects.event

import com.github.dedis.popstellar.model.objects.Meeting

/** This class is a builder for the Meeting object  */
class MeetingBuilder {
    private var id: String? = null
    private var name: String? = null
    private var creation: Long = 0
    private var start: Long = 0
    private var end: Long = 0
    private var location: String? = null
    private var lastModified: Long = 0
    private var modificationId: String? = null
    private var modificationSignatures: List<String>? = null

    constructor()
    constructor(meeting: Meeting) {
        id = meeting.id
        name = meeting.name
        creation = meeting.creation
        start = meeting.startTimestamp
        end = meeting.endTimestamp
        location = meeting.location
        lastModified = meeting.lastModified
        modificationId = meeting.modificationId
        modificationSignatures = meeting.modificationSignatures
    }

    fun setId(id: String?): MeetingBuilder {
        this.id = id
        return this
    }

    fun setName(name: String?): MeetingBuilder {
        this.name = name
        return this
    }

    fun setCreation(creation: Long): MeetingBuilder {
        this.creation = creation
        return this
    }

    fun setStart(start: Long): MeetingBuilder {
        this.start = start
        return this
    }

    fun setEnd(end: Long): MeetingBuilder {
        this.end = end
        return this
    }

    fun setLocation(location: String?): MeetingBuilder {
        requireNotNull(location) { "Location is null" }
        this.location = location
        return this
    }

    fun setLastModified(lastModified: Long): MeetingBuilder {
        this.lastModified = lastModified
        return this
    }

    fun setModificationId(modificationId: String?): MeetingBuilder {
        requireNotNull(modificationId) { "Modification Id is null" }
        this.modificationId = modificationId
        return this
    }

    fun setModificationSignatures(modificationSignatures: List<String>?): MeetingBuilder {
        requireNotNull(modificationSignatures) { "Modification signatures list is null" }
        this.modificationSignatures = modificationSignatures
        return this
    }

    fun build(): Meeting {
        checkNotNull(location) { "Location is null" }
        requireNotNull(modificationId) { "Modification Id is null" }
        requireNotNull(modificationSignatures) { "Modification signatures list is null" }
        return Meeting(
                id!!,
                name!!,
                creation,
                start,
                end,
                location!!,
                lastModified,
                modificationId!!,
                modificationSignatures)
    }
}